# SNPop
An application package for the collection, visualization and small scale analysis of catalogued data regarding single nucleotide polymorphismsacross populations.

Important: SNPanalysis utilizies the LDlinkR package, hence it requires a private user token to function. If you wish to use SNPanalysis replace the string: 'user_token' in the code, with a token retrieved from: https://ldlink.nci.nih.gov/?tab=apiaccess.


A has revolution occurred in the field of genomics. The human genome has become as accessible as ever and data regarding variations in the genome appear faster than they can be interpreted. For example, DbSNP, the most popular  database for single nucleotide variants, micro-satellites, and small-scale insertions and deletions,doubled its data size from  557,939,960 submissions to 907,237,763 between 2016-2017 and then again from 907,237,763 to 1,828,331,768 from 2017 to 2018 as its shown in the figure below. In 2021 DbSNP has 3,341,554,567 submissions. The evident acute growth in the amount of data available on human genetic variation as well as variation in other species classified genomics as a "big data" field and sparked the interest for the development of SNPop.






# SNPfinder.py



# Final Pipeline
The data flow/logic of SNPop's algorith is showcased in the figure below:
![1 vpd](https://user-images.githubusercontent.com/68243875/147920381-2196bc6b-e546-415a-9024-e4b58f510a98.png)

The data pipeline this algorithm produces is showcased in the following figure:
![pipeline](https://user-images.githubusercontent.com/68243875/147920326-039fb6c1-6b0e-4b80-ba94-c8259b55d29b.png)


